<h1 style= "color:green; text-align:center"> Web Mining and Applied NLP (44-620) </h1>

<h3 style= "color:orange"> Requests, JSON, and NLP</h3>

<p style= "color: orange; font-size: 18px"> Sudent Name: Curtis Rogers </p>

Perform the tasks described in the Markdown cells below.  When you have completed the assignment make sure your code cells have all been run (and have output beneath them) and ensure you have committed and pushed ALL of your changes to your assignment repository.

Make sure you have [installed spaCy and its pipeline](https://spacy.io/usage#quickstart) and [spaCyTextBlob](https://spacy.io/universe/project/spacy-textblob)

Every question that requires you to write code will have a code cell underneath it; you may either write your entire solution in that cell or write it in a python file (`.py`), then import and run the appropriate code to answer the question.

This assignment requires that you write additional files (either JSON or pickle files); make sure to submit those files in your repository as well.

In [2]:
# Create and activate a Python virtual environment. 
# Before starting the project, try all these imports FIRST
# Address any errors you get running this code cell 
# by installing the necessary packages into your active Python environment.
# Try to resolve issues using your materials and the web.
# If that doesn't work, ask for help in the discussion forums.
# You can't complete the exercises until you import these - start early! 
# We also import json and pickle (included in the Python Standard Library).

import json
import pickle

import requests
import spacy
from spacytextblob.spacytextblob import SpacyTextBlob

print('All prereqs installed.')
!pip list

All prereqs installed.
Package            Version
------------------ -----------
annotated-types    0.6.0
asttokens          2.4.1
blis               0.7.11
catalogue          2.0.10
certifi            2024.2.2
charset-normalizer 3.3.2
click              8.1.7
cloudpathlib       0.16.0
colorama           0.4.6
comm               0.2.2
confection         0.1.4
contourpy          1.2.0
cycler             0.12.1
cymem              2.0.8
debugpy            1.8.1
decorator          5.1.1
en-core-web-sm     3.7.1
executing          2.0.1
fonttools          4.50.0
idna               3.6
ipykernel          6.29.3
ipython            8.22.2
jedi               0.19.1
Jinja2             3.1.3
joblib             1.3.2
jupyter_client     8.6.1
jupyter_core       5.7.2
kiwisolver         1.4.5
langcodes          3.3.0
MarkupSafe         2.1.5
matplotlib         3.8.3
matplotlib-inline  0.1.6
mizani             0.11.0
murmurhash         1.0.10
nest-asyncio       1.6.0
nltk               3.8.1
numpy   

1. The following code accesses the [lyrics.ovh](https://lyricsovh.docs.apiary.io/#reference/0/lyrics-of-a-song/search) public api, searches for the lyrics of a song, and stores it in a dictionary object.  Write the resulting json to a file (either a JSON file or a pickle file; you choose). You will read in the contents of this file for future questions so we do not need to frequently access the API.

In [3]:
import requests
import json

AUTHOR='Henry David Thoreau'
POEM = 'The Summer Rain'

URL = f'https://poetrydb.org/author,title/{AUTHOR};{POEM}'
result = json.loads(requests.get(URL).text)
poem = '\n'.join(result[0]['lines']) 

print(poem)


My books I'd fain cast off, I cannot read,
'Twixt every page my thoughts go stray at large
Down in the meadow, where is richer feed,
And will not mind to hit their proper targe.
Plutarch was good, and so was Homer too,
Our Shakespeare's life were rich to live again,
What Plutarch read, that was not good nor true,
Nor Shakespeare's books, unless his books were men.

Here while I lie beneath this walnut bough,
What care I for the Greeks or for Troy town,
If juster battles are enacted now
Between the ants upon this hummock's crown?

Bid Homer wait till I the issue learn,
If red or black the gods will favor most,
Or yonder Ajax will the phalanx turn,
Struggling to heave some rock against the host.

Tell Shakespeare to attend some leisure hour,
For now I've business with this drop of dew,
And see you not, the clouds prepare a shower--
I'll meet him shortly when the sky is blue.

This bed of herd's grass and wild oats was spread
Last year with nicer skill than monarchs use.
A clover tuft is 

2. Read in the contents of your file.  Print the lyrics of the song (not the entire dictionary!) and use spaCyTextBlob to perform sentiment analysis on the lyrics.  Print the polarity score of the sentiment analysis.  Given that the range of the polarity score is `[-1.0,1.0]` which corresponds to how positive or negative the text in question is, do you think the lyrics have a more positive or negative connotaion?  Answer this question in a comment in your code cell.

In [4]:
#I did the imports at the beginning

#Loads English Tokenizer
nlp = spacy.load('en_core_web_sm')
nlp.add_pipe('spacytextblob')

#Document
doc = nlp(poem)

#Sentiment analysis

polarity_score = doc._.polarity

print(doc)
print()
print('*****************************************************************')
print(f'the polarity score is {polarity_score}, Since this score is {polarity_score} I would say that the poem has a neutral connotation')



My books I'd fain cast off, I cannot read,
'Twixt every page my thoughts go stray at large
Down in the meadow, where is richer feed,
And will not mind to hit their proper targe.
Plutarch was good, and so was Homer too,
Our Shakespeare's life were rich to live again,
What Plutarch read, that was not good nor true,
Nor Shakespeare's books, unless his books were men.

Here while I lie beneath this walnut bough,
What care I for the Greeks or for Troy town,
If juster battles are enacted now
Between the ants upon this hummock's crown?

Bid Homer wait till I the issue learn,
If red or black the gods will favor most,
Or yonder Ajax will the phalanx turn,
Struggling to heave some rock against the host.

Tell Shakespeare to attend some leisure hour,
For now I've business with this drop of dew,
And see you not, the clouds prepare a shower--
I'll meet him shortly when the sky is blue.

This bed of herd's grass and wild oats was spread
Last year with nicer skill than monarchs use.
A clover tuft is 

<h2 style= color:green> Question 2 cont.</h2>

<p> The polarity score for this poem was very close to neutral, falling just slightly on the positive side. I feel that this is an accurate representation for this poem. </p>

3. Write a function that takes an artist, song, and filename, accesses the lyrics.ovh api to get the song lyrics, and writes the results to the specified filename.  Test this function by getting the lyrics to any four songs of your choice and storing them in different files.

In [11]:
 #this will get the poem if given the author and title. Then save it to a certain filename
def fetch_poem(author,title,filename):
    URL = f'https://poetrydb.org/author,title/{author};{title}'
    data = json.loads(requests.get(URL).text)
    with open(filename, 'w') as file:
        json.dump(data, file)
    print(f'Saved file {filename}')



#fetch poem

fetch_poem('William Shakespear', 'Spring', 'Shakespear_Spring')
fetch_poem('Ralph Waldo Emerson', 'Ode to Beauty', 'Emerson_OdeToBeauty')
fetch_poem('Edgar Allan Poe', 'The Sleeper', 'Poe_TheSleeper')
fetch_poem('Edgar Allan Poe', 'The City in The Sea', 'Poe_TheCityInTheSea')


Saved file Shakespear_Spring
Saved file Emerson_OdeToBeauty
Saved file Poe_TheSleeper
Saved file Poe_TheCityInTheSea


4. Write a function that takes the name of a file that contains song lyrics, loads the file, performs sentiment analysis, and returns the polarity score.  Use this function to print the polarity scores (with the name of the song) of the three files you created in question 3.  Does the reported polarity match your understanding of the song's lyrics? Why or why not do you think that might be?  Answer the questions in either a comment in the code cell or a markdown cell under the code cell.

In [14]:
def poem_sentiment (filename):
    with open(filename, 'r') as file:
        data = json.load(file)
        poem = '\n'.join(data[0]['lines'])
    #Loads English Tokenizer
    nlp = spacy.load('en_core_web_sm')
    nlp.add_pipe('spacytextblob')

    #Document
    doc = nlp(poem)

    #Sentiment analysis
    polarity_score = doc._.polarity

    print()
    print('*****************************************************************')
    print(f'the polarity score for{filename} is {polarity_score}')

#new poem sentiment analysis

poem_sentiment('Emerson_OdeToBeauty')
poem_sentiment('Poe_TheCityInTheSea')
poem_sentiment('Poe_TheSleeper')
poem_sentiment('Shakespear_Spring')
    


*****************************************************************
the polarity score forEmerson_OdeToBeauty is 0.1265627705627706

*****************************************************************
the polarity score forPoe_TheCityInTheSea is -0.01367337164750958

*****************************************************************
the polarity score forPoe_TheSleeper is 0.13499007936507934

*****************************************************************
the polarity score forShakespear_Spring is -0.14423076923076925


<h2 style= color:green> Question 4 cont.</h2>

|Title|Author|Polarity|
|-----|------|--------|
|"Ode To Beauty"|Emerson|0.1266|
|"The City in The Sea"|Poe|-0.0137|
|"The Sleeper"|Poe|0.1351|
|"Spring"|Shakespear|-0.1442|

<p>Most of the polarity scores for the poems that I chose where fairly neutral. The results had 2 poems leaning slightly negative and 2 leaning slightly positive. For the most part I agree with all of the results except Shakespears poem "Spring". In my opinion this poem should be on the slightly positive side. I think that this sentiment was incorrect for this poem because it contained the words "fear" and "unpleasing". These words are normally interpreted with a negative sentiment, and might be what lead to the slight negative polarity result with this poem.

Other than that I think that the other were pretty accurate. I am no poetry expert. So there could be some error with my interpretation aswell. </p>

In [15]:
!jupyter nbconvert --to html requests-json-nlp.ipynb

[NbConvertApp] Converting notebook requests-json-nlp.ipynb to html
[NbConvertApp] Writing 297631 bytes to requests-json-nlp.html
